### DISCUSSION TITLE
Exclude files in `/api/` from being registered as serverless functions

### DISCUSSION THREAD

**Discussion by sachaw** [vercel/community#46] [Help] [answered]
Hi, I have a monorepo setup (https://github.com/sachaw/scrobble-moe) in `/server/api` there exists `app.ts` and `webhook.ts` these are the only two serverless functions that I want to run, the rest of the files `utils/*` and `graphql/*` should not be serverless functions.
How can I configure to either ignore or specify only the specific functions that I want to register, I have tried several configurations of `vercel.json` however none have seemed to work

**Comment by mcsdevv** [comment, marked as answer]
Hello @sachaw, thanks for raising this!

Rather than excluding files or directories, we recommend using a directory outside of `/api` for any files that should not be created as Serverless Functions.

**Reply by agusterodin** [reply]
What if we want to have frontend api logic (for example, our RTK query setup) in a folder called "api"? There is no reason why that shouldn't be allowed.

**Reply by mcsdevv** [reply]
Hi @agusterodin, this is a limitation with the way it is implemented currently. That's not to say it can't be changed though and is certainly something that can be considered for the future.

**Reply by agusterodin** [reply]
Thank you!

**Reply by AaronBeaudoin** [reply]
Just want to add my two cents: **This is a really poor design choice on Vercel's part.** If any framework or project uses the `api/` directory for something other than what Vercel wants it to be used for, the user then has to find some sort of hack to get around their projects structure in order to make Vercel happy.

Vercel needs to add a configuration option allowing functions to be placed in a directory of the user's choice. Forcing the user to conform to an exact directory structure may seem nice for consistency, but it is always bound to cause problems for some users. Really the `vercel.json` config file should just allow defined functions to have a user configurable entry point and the key of the configuration object should be the "output URL" where the function will be served. That way, users can do something like having a function at `vercel/render.js` in their project and then serve it at `/` in production.

**Comment by JannikZed** [comment]
we are having the same issue using a monorepo with strapi and trying to build and deploy the admin dashboard separately as vercel project ..

**Comment by sampoder** [comment]
For anyone wondering check out: https://vercel.com/guides/prevent-uploading-sourcepaths-with-vercelignore

**Comment by cprcrack** [comment]
On my Node.js, typescript API working on a legacy Express.js app (not sure if applicable to other runtimes), I ended doing the following:
1. Rename the `/api/` folder to `/api-src/`, where you can keep all the files you want without worrying about Vercel converting them to API endpoints. In my case my main entry-point is now `/api-src/api.ts`. 
2. Create an empty `/api/` folder.
3. Create an `/api/index.ts` single file with the following content:

```
import '../api-src/api';
```
4. In order for dependencies to be installed and resolved once deployed, create the following unix-like symlinks from `/` like this:

```
ln -s /api-src/package.json package.json
ln -s /api-src/package-lock.json package-lock.json
```
(You could also just move or copy the files from `/api-src/` to `/` if that works in your development environment)

5. Finally I kept the `/vercel.json` file as it was before:
```
{
  "rewrites": [{ "source": "/api/(.*)", "destination": "/api" }]
}
```
